src/NLog/Targets/TargetWithContext.cs

```diff
@@ -276,6 +277,175 @@ protected bool ShouldIncludeProperties(LogEventInfo logEvent)
+            private IEnumerable<KeyValuePair<string, object>> _scopeContextProperties;
+            private ICollection<string> _gdcNames;
+
+            private bool _CountPropertiesCalled;
+            private int _propertiesCount;
+            private bool _checkExcludeProperties;
+
+            private KeyValuePair<string, object>? BuildPair(string propertyName, object propertyValue)
+            {
+                if (string.IsNullOrEmpty(propertyName))
+                    return null;
+
+                if (_checkExcludeProperties && _targetContext.ExcludeProperties.Contains(propertyName))
+                    return null;
+
+                return new KeyValuePair<string, object>(propertyName, propertyValue);
+            }
+
+            private KeyValuePair<string, object>? BuildPairSerialized(string propertyName, object propertyValue)
+            {
+                if (string.IsNullOrEmpty(propertyName))
+                    return null;
+
+                if (_checkExcludeProperties && _targetContext.ExcludeProperties.Contains(propertyName))
+                    return null;
+
+                if (!_targetContext.SerializeScopeContextProperty(_logEvent, propertyName, propertyValue, out var serializedValue))
+                    return null;
+
+                return new KeyValuePair<string, object>(propertyName, serializedValue);
+            }
+
+            public int PropertiesCount
+            {
+                get => _propertiesCount;
+            }
+
+            private GetAllPropertiesCountContext() { }
+
+            public GetAllPropertiesCountContext(TargetWithContext targetContext, LogEventInfo logEvent)
+            {
+                if (targetContext is null)
+                    throw new ArgumentNullException(nameof(targetContext));
+                if (logEvent is null)
+                    throw new ArgumentNullException(nameof(logEvent));
+
+                _targetContext = targetContext;
+                _logEvent = logEvent;
+            }
+
+            public GetAllPropertiesCountContext EstimateCount()
+            {
+                if (_CountPropertiesCalled)
+                    throw new InvalidOperationException("The method should be called only once!");
+                _CountPropertiesCalled = true;
+
+                _propertiesCount += _targetContext.ContextProperties?.Count ?? 0;
+
+                if (_targetContext.IncludeEventProperties && _logEvent.HasProperties)
+                {
+                    _logEventProperties = _logEvent.Properties;
+                    _propertiesCount += _logEventProperties.Count;
+                }
+
+                if (_targetContext.IncludeScopeProperties)
+                {
+                    IDictionary<string, object> combinedProperties = null;
+                    var contextLayout = _targetContext._contextLayout.ScopeContextPropertiesLayout;
+                    if (!_targetContext.CombineProperties(_logEvent, contextLayout, ref combinedProperties))
```

Comment:
When `CombineProperties` returns `true` and have assigned `ref combinedProperties` then it should be used as `_scopeContextProperties`.

Reply:
Sorry, not sure if I followed you quite well... Do you mean something like that?

```
if (_targetContext.IncludeScopeProperties)
{
    IDictionary<string, object> combinedProperties = null;
    var contextLayout = _targetContext._contextLayout.ScopeContextPropertiesLayout;
    if (!_targetContext.CombineProperties(_logEvent, contextLayout, ref combinedProperties))
    {
        _scopeContextProperties = ScopeContext.GetAllProperties();
        _propertiesCount += _scopeContextProperties.Count();
    }
    // NEW
    else if(combinedProperties?.Count > 0)
    {
        _scopeContextProperties = combinedProperties;
        _propertiesCount += combinedProperties.Count;
    }
}

```

​Assuming it is true, I made my latest tentative commit

Reply:
Again the goal is to reduce allocation at all costs, so adding `ToDictionary` is not optimal. The same with the allocation of the new private class.

But now the unit-test are green, so maybe I can merge, and adjust the code myself afterwards.

Reply:
Thank you for your patience